### ISSUE TITLE
Is Android supported?

### ISSUE THREAD

**Issue by allidev** [HowardHinnant/date#161]
When I tried to build date on Android, it reported a fatal error: 
tz.cpp:128:23: fatal error: wordexp.h: No such file or directory

Does date work on Android?
Thanks!

**Comment by HowardHinnant** [comment]
The intent is to support Android.  But to be honest, this is a one-pony-show and I'm the only pony, and without an Android device to test on.  `<wordexp.h>` is an external utility to expand path shortcuts such as "~", and is not a major dependence on this library.  This is just used to locate the full file path of the downloaded IANA tz database.

I would be very grateful for any workarounds you might be able to supply to locate an alternate default location for the downloaded IANA tz database.  For example on `iOS`, `date::iOSUtils::get_tzdata_path()` is called instead.  I have no idea what this does, as someone else wrote it, but it boils down to providing an alternative `std::string` path to where the IANA database lives.

Once we get past such fundamentals as how the IANA database is located, this library is very portable, and should work on Android (assuming it has `<chrono>` support).

**Comment by HowardHinnant** [comment]
It might be that I should not attempt to `#include <wordexp.h>` if `USE_OS_TZDB == 1`.  This is a theory which has not been tested.  However `USE_OS_TZDB==1` means that the one should use the OS-supplied database,  in which case the library should know where that is, and does not need the aid of `<wordexp.h>`.

I.e. `#include <wordexp.h>` is likely a gratuitous (unnecessary) include when `USE_OS_TZDB == 1`.

**Comment by allidev** [comment]
I am working on a project what can benefit from date library on Android. I will take a crack at this and see what I come up with.

**Comment by HowardHinnant** [comment]
Commit https://github.com/HowardHinnant/date/commit/c5e58fd0156c068b1135fbd275e92d4bb72f63f6 removes the dependence on `<wordexp.h>` when tz.cpp is compile with `-DUSE_OS_TZDB`.

**Comment by hjmallon** [comment]
Bionic is the c library used on andorid. Here is the function which seems to look for the android tzdata.

https://github.com/AOKP/bionic/blob/55495fbdc0ef24212fd108153d784070cd15370d/libc/tzcode/bionic.cpp#L198

**Comment by HowardHinnant** [comment]
Will reopen if this is still a problem _and_ someone is willing to work on it.

**Comment by molind** [comment]
Hi,

Faced the same problem on Android. And solved it with smallest possible fix. I believe it won't break anything. :)

Preface: we build our app and valhalla as one of it's dependencies, on iOS and Android, both use date.h inside. iOS build's without issues with command line: 

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -std=c++14 -DNDEBUG=1 -DUSE_STD_REGEX=1 -DRAPIDJSON_HAS_STDSTRING=1 -DPACKAGE_VERSION="\"2.7.0\"" -I. -Ivalhalla -Igenfiles -Ithird_party/rapidjson/include -Ithird_party/date/include -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.0.sdk -miphoneos-version-min=8.0 -I/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.0.sdk/usr/include -I/Users/evgen/git/libs/build/iOS/x86_64/include -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.0.sdk -miphoneos-version-min=8.0 -O3 -fembed-bitcode -fvisibility=hidden -std=c++14 -stdlib=libc++ -fno-aligned-allocation -DHAS_REMOTE_API=0 -DUSE_OS_TZDB=1 -DMISSING_LEAP_SECONDS=0 -c third_party/date/src/tz.cpp -o third_party/date/src/tz.o
```

while Android can't find *wordexp.h*

```
armv7a-linux-androideabi16-clang++ -std=c++14 -DNDEBUG=1 -DUSE_STD_REGEX=1 -DRAPIDJSON_HAS_STDSTRING=1 -DPACKAGE_VERSION="\"2.7.0\"" -I. -Ivalhalla -Igenfiles -Ithird_party/rapidjson/include -Ithird_party/date/include -I/Users/evgen/git/libs/build/Android/armeabi-v7a/include -fPIC -O3 -std=c++17 -march=armv7-a -mfloat-abi=softfp -DHAS_REMOTE_API=0 -DUSE_OS_TZDB=0 -DINSTALL=. -DMISSING_LEAP_SECONDS=0 -c third_party/date/src/tz.cpp -o third_party/date/src/tz.o
third_party/date/src/tz.cpp:145:14: fatal error: 'wordexp.h' file not found
#    include <wordexp.h>
             ^~~~~~~~~~~
1 error generated.
make[3]: *** [third_party/date/src/tz.o] Error 1
```

I checked do we really need wordexp.h in our configuration and when it's used?
It's used in expand_path(std::string path) function only when INSTALL macro is not defined and !USE_OS_TZDB.

```
#    if !defined(INSTALL)

static
std::string
expand_path(std::string path)
```

While `wordexp.h` check in beginning checks `#  if !USE_OS_TZDB`. I propose to change check in beginning of file to the same rule an it works for function:
```diff
diff --git a/src/tz.cpp b/src/tz.cpp
index 38e14bc..78ab35f 100644
--- a/src/tz.cpp
+++ b/src/tz.cpp
@@ -141,7 +141,7 @@
 #  endif  // HAS_REMOTE_API
 #else   // !_WIN32
 #  include <unistd.h>
-#  if !USE_OS_TZDB
+#  if !USE_OS_TZDB && !defined(INSTALL)
 #    include <wordexp.h>
 #  endif
 #  include <limits.h>
```

After this fix it's possible to build date.h on Android and iOS and it works without issues.

**Comment by TheStormN** [comment]
Hello @HowardHinnant 

Although the the library builds on Android when `USE_OS_TZDB` is defined, no time zones are available from the library, the list is empty, you can neither get the current time zone or any other one by name.

As I'm not an Android developer, just working on a codebase which is used on many platforms, so I've just installed Android Studio and created a simple C++ application and started debugging using the built-in emulator and I've found 3 issues.
TLDR; the third issue is the real showstopper.

- Fetching the current time zone(string) should be done via `__system_property_get("persist.sys.timezone", sys_timezone);`. I added that and it is working perfectly, so the first issue is and easy fix.
- Second issue is the location of the TZ database. I used the hint from @hjmallon about fetching the location, but I've used the newer version of the bionic.cpp which does use `ANDROID_TZDATA_ROOT` env variable to fetch the path and of course if that fails it checks for older paths too. So that was an easy fix too. I've tested it and it is working good.
- The third issue however is more serious and cannot be fixed with simple patches. On Android the TZ database is stored into a single binary file called `tzdata`. That file contains everything and the format itself is called `NodaZoneData`. There is a compiler which is using the official IANA files to generate that file. However the `date` library have no code which can handle that.

So, I know that you also don't have Android device to develop with, but using Android Studio with the built-in emulator is good enough, you can even switch between different versions of Android. Perhaps when and if you have time you can research a bit the binary format and make a really simple parser which can fetch the data? Thanks :)
